lib/socket.js

```diff
@@ -156,45 +158,42 @@ module.exports = class UDXSocket extends events.EventEmitter {
     return binding.udx_napi_socket_send_buffer_size(this._handle, size)
   }
 
-  send (buffer, offset, length, port, ip, ttl, onflush) {
-    if (typeof ttl === 'function') {
-      onflush = ttl
-      ttl = 0
-    }
+  async send (buffer, port, host, ttl) {
+    // TODO: should this be an error?
+    if (this.closing) throw new Error('Socket is closed')
 
-    if (this.closing) {
-      if (onflush) onflush(new Error('Socket is closed'))
-      return
-    }
+    if (!host) host = '127.0.0.1'
 
     if (!this.bound) this.bind(0)
 
     const id = this._allocSend()
     const req = this._reqs[id]
 
-    req.buffer = buffer.subarray(offset, offset + length)
-    req.onflush = onflush || noop
+    req.buffer = buffer
 
-    binding.udx_napi_socket_send_ttl(this._handle, req.handle, id, buffer, port, ip, ttl || 0)
-  }
+    const promise = new Promise((resolve, reject) => {
+      req.resolve = resolve
+      req.reject = reject
+    })
 
-  _allocSend () {
-    if (this._free.length > 0) return this._free.pop()
-    const handle = b4a.allocUnsafe(binding.sizeof_udx_socket_send_t)
-    return this._reqs.push({ handle, buffer: null, onflush: null }) - 1
+    binding.udx_napi_socket_send_ttl(this._handle, req.handle, id, buffer, port, host, ttl || 0)
+
+    return promise
   }
 
-  static createSocket () {
-    return new this()
+  trySend (buffer, port, host, ttl) {
+    this.send(buffer, port, host, ttl).catch(noop)
```

Comment:
at some point, it would be nice if trySend could avoid allocating a promise all together to help with gc pressure